### PR TITLE
Reorganize libs.native

### LINKED ISSUE #61864: Move libraries shims under src/native
The current directory structure under src/libraries/native is confusing:
- The sub-directory names do not match the OSes where the shims are actually building. 
- The shims are statically linked with the runtime in some configurations 

We should consider moving all shims under src/native that seems to be the best place for unmanaged code shared between runtimes.

See https://github.com/dotnet/runtime/issues/61852#issuecomment-974545733 for full context

### PR DESCRIPTION
**Pull request by am11**
Fixes https://github.com/dotnet/runtime/issues/61864.

### CODE REVIEW

**Comment by am11** [comment]
@jkotas, @carlossanlop, with this structure, there is no Unix, Windows and AnyOS subdirectories anymore and `System.IO.Compression/` subdir has everything required by that lib. The merged cmake configs will make things relatively easier if/when zlib build from source on Unix is enabled.

@safern, @jkoritzinsky, I've moved `native-binplace.proj` to `src/libraries/native-binplace.proj` instead; since it relies on common libraries' props/targets.

File: eng/pipelines/common/evaluate-default-paths.yml [hunk 1]
```diff
@@ -9,8 +9,8 @@ jobs:
     - subset: coreclr
       include:
       - src/libraries/System.Private.CoreLib/*
-      - src/libraries/Native/Unix/System.Globalization.Native/*
-      - src/libraries/Native/Unix/Common/*
+      - src/native/libs/System.Globalization.Native/*
+      - src/native/libs/Common/*
```
Comment (jkotas) [inline eng/pipelines/common/evaluate-default-paths.yml:13]:
```suggestion
```
This should be covered by `src/native/*`

File: eng/pipelines/common/evaluate-default-paths.yml [hunk 2]
```diff
@@ -29,8 +29,8 @@ jobs:
     - subset: mono
       include:
       - src/libraries/System.Private.CoreLib/*
-      - src/libraries/Native/Unix/System.Globalization.Native/*
-      - src/libraries/Native/Unix/Common/*
+      - src/native/libs/System.Globalization.Native/*
+      - src/native/libs/Common/*
```
Comment (jkotas) [inline eng/pipelines/common/evaluate-default-paths.yml:33]:
```suggestion
```
This should be covered by `src/native/*`

**Review by jkotas — APPROVED** [review]
Looks great! Thank you for doing this.

I was hoping that the shims can be directly in the `native` directory, but I understand why you have put them `native/libs` and I do not have a better suggestion. So LGTM.

File: eng/native/configurecompiler.cmake [hunk 3]
```diff
@@ -525,7 +525,7 @@ if (MSVC)
 
   # disable C++ RTTI
   # /GR is added by default by CMake, so remove it manually.
-  string(REPLACE "/GR" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
+  string(REPLACE "/GR " " " CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
```
Comment (carlossanlop) [inline eng/native/configurecompiler.cmake:528]: Intentional? Is this change needed?

File: eng/pipelines/common/evaluate-default-paths.yml [hunk 4]
```diff
@@ -9,8 +9,6 @@ jobs:
     - subset: coreclr
       include:
       - src/libraries/System.Private.CoreLib/*
-      - src/libraries/Native/Unix/System.Globalization.Native/*
```
Comment (carlossanlop) [inline eng/pipelines/common/evaluate-default-paths.yml:12]: cc @safern - you mentioned Globalization in the issue I recently opened, asking about our zlib consumption.

**Review by carlossanlop — APPROVED** [review]
@am11  @jkotas - For my own education, can you please confirm if I'm understanding the changes in this PR correctly? Please correct me if I got something wrong:

All the native code is built using the same cmake script(s), but depending on the OS, we choose which code files to build for that OS:

  - Th `pal_*.c|.h` files contain syscalls that only apply to Unix and WASM..
  - The `zlib` stuff is cross-platform, and could be built in any OS, but we currently only consume it in Windows, but maybe in the future it could be consumed by others.
  - The new structure will be clearer and will avoid future confusion like when I recently asked why zlib lived inside the Windows folder.

At first glance the changes look good to me, but I'm far from being the expert in cmake stuff, so I'll defer the decision to merge to @jkotas.

File: eng/native/configurecompiler.cmake (hunk 3 above)
Comment (am11) [reply eng/native/configurecompiler.cmake:528]: Yup, it is ported from old src/libraries/Native/Windows/CMakeFiles.txt. Without it, it replaces `/GR-` (set by default in VS2022) with `-` and that gives (non-fatal) warnings.

**Comment by am11** [comment]
> All the native code is built using the same cmake script(s), but depending on the OS, we choose which code files to build for that OS:

That's right. The cmake configuration are now unified, to make it easier to enable which sources to compile without changing the directory structure. This simplifies things; no more indirections and conditions outside the cmake script.

> Th `pal_*.c|.h` files contain syscalls that only apply to Unix and WASM..

Currently these utility headers are only used by Unix. When a source file which compiles on Windows require them, we can sprinkle around some `ifdef`s to make it compile. :)

> The zlib stuff is cross-platform, and could be built in any OS, but we currently only consume it in Windows, but maybe in the future it could be consumed by others.

Yup, enabling zlib compile-from-src will be a matter of hoisting these lists https://github.com/dotnet/runtime/blob/92854b914abe1ca2796ffc6c0abf3a076b222313/src/native/libs/System.IO.Compression.Native/CMakeLists.txt#L115
out of `if (CLR_CMAKE_TARGET_UNIX OR CLR_CMAKE_TARGET_BROWSER) / else ()` condition and adding ZLIB_SOURCES to Unix.
On that note, we still have three copies of zlib in the repo:
* src/mono/mono/zlib
* src/native/libs/System.IO.Compression.Native/zlib-intel
* src/native/libs/System.IO.Compression.Native/zlib

it might be possible to converge to one (less) flavor. (cc @lambdageek, @akoeplinger)

File: eng/pipelines/common/evaluate-default-paths.yml (hunk 4 above)
Comment (safern) [reply eng/pipelines/common/evaluate-default-paths.yml:12]: This is just pure cleanup as this will now be covered by `-src/native/*`?

File: eng/pipelines/common/evaluate-default-paths.yml [hunk 5]
```diff
@@ -9,8 +9,6 @@ jobs:
...
-      - src/libraries/Native/Unix/Common/*
       - src/native/*
```
Comment (safern) [inline eng/pipelines/common/evaluate-default-paths.yml:12]: should we exclude from the coreclr path partition the shims that are not affecting/linked into the runtime? i.e: 
exclude: `src/native/libs/*` but include :`src/native/System.Globalization.Native/*` and whatever common paths are shared for `Globalization.Native?`

File: eng/pipelines/common/evaluate-default-paths.yml [hunk 6]
```diff
@@ -29,8 +27,6 @@ jobs:
     - subset: mono
       include:
       - src/libraries/System.Private.CoreLib/*
-      - src/libraries/Native/Unix/System.Globalization.Native/*
```
Comment (safern) [inline eng/pipelines/common/evaluate-default-paths.yml:32]: Same here.

File: eng/pipelines/common/evaluate-default-paths.yml [hunk 7]
```diff
@@ -29,8 +27,6 @@ jobs:
...
-      - src/libraries/Native/Unix/Common/*
```
Comment (safern) [inline eng/pipelines/common/evaluate-default-paths.yml:33]: installer path partition should exclude `src/native/libs/*`?

File: eng/pipelines/common/evaluate-default-paths.yml (hunk 4 above)
Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:12]: Yup, https://github.com/dotnet/runtime/pull/61958#discussion_r755328194, we previously added `src/native/*` path in https://github.com/dotnet/runtime/commit/565f3ee5ad68dad35305e9157e37fcdccc2aa2b6 (to trigger build of all related subsets when something changes under this directory).

File: eng/pipelines/common/evaluate-default-paths.yml (hunk 7 above)
Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:33]: coreclr builds static host. Technically, it can be affected by changes in libs.native (by means of entrypoints.h, which p/invoke dispatcher relies on).

File: eng/pipelines/common/evaluate-default-paths.yml (hunk 5 above)
Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:12]: The depndencies in native libs and other libs code is not that striaightforward. CoreCLR build parts of libs as well as corehost, the latter then relies on parts of libs. I wonder if it is better to build all related subsets when something changes in src/native/*? cc @VSadov

**Review by safern — APPROVED** [review]
Other than some adjustments on the evaluate path subsets, looks good to me.

File: eng/pipelines/common/evaluate-default-paths.yml (hunk 7 above)
Comment (safern) [reply eng/pipelines/common/evaluate-default-paths.yml:33]: Well, if changes under `src/libraries/Native` happened before this change, the installer subset was not tested, so I think that's fine.

Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:33]: @safern, this comment is on mono subset, did you mean to point out mono or installer?

Comment (safern) [reply eng/pipelines/common/evaluate-default-paths.yml:33]: I meant installer, I can't comment on installer since the GH diff won't let me since it wasn't changed.

Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:33]: installer does not include `src/native/libs/*`.

**Comment by carlossanlop** [comment]
> On that note, we still have three copies of zlib in the repo:
> * src/mono/mono/zlib
> * src/native/libs/System.IO.Compression.Native/zlib-intel
> * src/native/libs/System.IO.Compression.Native/zlib
> 
> it might be possible to converge to one (less) flavor.

Oh I didn't know there was also a separate zlib in mono. I'll share this info in my PR: https://github.com/dotnet/runtime/pull/61883

File: eng/pipelines/common/evaluate-default-paths.yml [hunk 8]
```diff
@@ -47,6 +43,9 @@ jobs:
       - eng/pipelines/coreclr/*
       - eng/pipelines/libraries/*
     - subset: libraries
+      include:
```
Comment (safern) [inline eng/pipelines/common/evaluate-default-paths.yml:46]: No need to add these. When no `include` is specified it means it includes everything, but the specified subsets on the `exclude` parameter. Which in this case, from `src/native` it only excludes `src/native/eventpipe/*`

File: eng/pipelines/common/evaluate-default-paths.yml [hunk 9]
```diff
@@ -68,6 +67,8 @@ jobs:
     - subset: installer
       include:
       - docs/manpages/*
+      - src/native/common/*
```
Comment (safern) [inline eng/pipelines/common/evaluate-default-paths.yml:70]: This is not needed either. You can see this comment to see how the tooling fully works: https://github.com/dotnet/runtime/blob/main/eng/pipelines/evaluate-changed-paths.sh#L2-L17

File: eng/pipelines/common/evaluate-default-paths.yml (hunk 7 above)
Comment (safern) [reply eng/pipelines/common/evaluate-default-paths.yml:33]: It does, look at my other comments. Basically installer is defined as this:

```yml
    - subset: installer
      include:
      - docs/manpages/*
      exclude:
      - eng/Version.Details.xml
      - '*.md'
      - LICENSE.TXT
      - PATENTS.TXT
      - THIRD-PARTY-NOTICES.TXT
      - docs/*
      - src/coreclr/*
      - src/mono/*
      - src/libraries/*
      - src/tests/*
      - src/native/eventpipe/*
      - eng/pipelines/coreclr/*
      - eng/pipelines/mono/*
      - eng/pipelines/libraries/*
```

Which means, include everything but what's excluded in that list + `docs/manpages/*` the reason why `docs/manpages/*` needs to be in the `include` list is because we explicitly exclude: `docs/*`. 

So that means that it will include everything under `src/native/*` except `src/native/eventpipe/*` so that is why I suggest adding and exclusion to this list to exclude `src/native/libs/*`

File: eng/pipelines/common/evaluate-default-paths.yml (hunk 8 above)
Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:46]: OK, does libraries subset include `src/native` implicitly somehow and coreclr / mono don't?

File: eng/pipelines/common/evaluate-default-paths.yml (hunk 7 above)
Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:33]: Ah, gotcha. Just so you know, this code refers to libs.native stuff:
```
src/native/corehost/apphost/static/CMakeLists.txt
src/native/corehost/hostpolicy/hostpolicy_context.cpp
```
I understand that you are saying that it's fine to ignore building corehost when stuff changes in libs.

File: eng/pipelines/common/evaluate-default-paths.yml (hunk 8 above)
Comment (safern) [reply eng/pipelines/common/evaluate-default-paths.yml:46]: Yes, because there is no `include:` paths specified for libraries. This is how the mechanism works:

```
Scenarios:
  1. exclude paths are specified
      Will include all paths except the ones in the exclude list.
  2. include paths are specified
      Will only include paths specified in the list.
  3. exclude + include:
      1st we evaluate changes for all paths except ones in excluded list. If we can not find
      any applicable changes like that, then we evaluate changes for incldued paths
      if any of these two finds changes, then a variable will be set to true.
  In order to consume this variable in a yaml pipeline, reference it via: $[ dependencies.<JobName>.outputs["<StepName>_<subset>.containschange"] ]
  Example:
  -difftarget ''HEAD^1'' -subset coreclr -includepaths ''src/libraries/System.Private.CoreLib/*'' -excludepaths ''src/libraries/*+src/installer/*''
  This example will include ALL path changes except the ones under src/libraries/*!System.Private.CoreLib/*
```

Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:46]: Yup, just read your other comments. Thanks for the link.

Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:46]: We made this change previously because coreclr and mono builds weren't running when `src/native/eventpipe` changed: https://github.com/dotnet/runtime/commit/565f3ee5ad68dad35305e9157e37fcdccc2aa2b6. Looks like point 3 didn't hold there for coreclr and mono since exclude list doesn't have src/native?

Comment (am11) [reply eng/pipelines/common/evaluate-default-paths.yml:46]: Reading this comment (from parent template): https://github.com/dotnet/runtime/blob/c158cb2fe35c17d8fb7de802c1ffba74c2b936b2/eng/pipelines/common/evaluate-paths-job.yml#L16-L24
looks like "all paths" refers to all paths of that subset, not globally?

Comment (safern) [reply eng/pipelines/common/evaluate-default-paths.yml]: That comment is wrong apparently it meant: 

```
This example will exclude ALL path changes under src/libraries/ except the ones under src/libraries/*!System.Private.CoreLib/* 
```

`include -> exclude`.

Comment (safern) [reply eng/pipelines/common/evaluate-default-paths.yml]: > Looks like point 3 didn't hold there for coreclr and mono since exclude list doesn't have src/native?

Yeah, weird. Now I'm curious why. Let me run a test real quick.

**Comment by safern** [comment]
@am11 I was curious at what we saw previously that we needed to add `src/native/*` to the mono and coreclr subsets on the change you pointed out and it seems like it was a confusion, what was missing was to include the condition on the `runtime_tests` changed path variable? I just ran this test with what you have at the moment for coreclr subset (removing `src/native/*` from the include list:

```diff
>> git diff
@@ -1,6 +1,6 @@
 #ifndef __DIAGNOSTICS_DUMP_PROTOCOL_H__
 #define __DIAGNOSTICS_DUMP_PROTOCOL_H__
-
+// change
 #include "ds-rt-config.h"
 
 #ifdef ENABLE_PERFTRACING
```

```bash
>> eng/pipelines/evaluate-changed-paths.sh --azurevariable containsChange --difftarget dotnet/main --subset coreclr --includepaths 'src/libraries/System.Private.CoreLib/*+src/native/libs/Common/*+src/native/libs/System.Globalization.Native/*+src/native/libs/System.IO.Compression.Native/*' --excludepaths 'eng/Version.Details.xml+*.md+LICENSE.TXT+PATENTS.TXT+THIRD-PARTY-NOTICES.TXT+docs/*+src/installer/*+src/mono/*+src/libraries/*+src/native/libs/*+src/tests/*+eng/pipelines/installer/*+eng/pipelines/mono/*+eng/pipelines/libraries/*'


******* Probing coreclr exclude paths *******
eng/Version.Details.xml
*.md
LICENSE.TXT
PATENTS.TXT
THIRD-PARTY-NOTICES.TXT
docs/*
src/installer/*
src/mono/*
src/libraries/*
src/native/libs/*
src/tests/*
eng/pipelines/installer/*
eng/pipelines/mono/*
eng/pipelines/libraries/*

+ git diff -M -C -b --ignore-cr-at-eol --ignore-space-at-eol --exit-code --quiet dotnet/main -- ':!eng/Version.Details.xml' ':!*.md' ':!LICENSE.TXT' ':!PATENTS.TXT' ':!THIRD-PARTY-NOTICES.TXT' ':!docs/*' ':!src/installer/*' ':!src/mono/*' ':!src/libraries/*' ':!src/native/libs/*' ':!src/tests/*' ':!eng/pipelines/installer/*' ':!eng/pipelines/mono/*' ':!eng/pipelines/libraries/*'

----- Matching files for coreclr -----
+ git diff -M -C -b --ignore-cr-at-eol --ignore-space-at-eol --name-only dotnet/main -- ':!eng/Version.Details.xml' ':!*.md' ':!LICENSE.TXT' ':!PATENTS.TXT' ':!THIRD-PARTY-NOTICES.TXT' ':!docs/*' ':!src/installer/*' ':!src/mono/*' ':!src/libraries/*' ':!src/native/libs/*' ':!src/tests/*' ':!eng/pipelines/installer/*' ':!eng/pipelines/mono/*' ':!eng/pipelines/libraries/*'
src/native/eventpipe/ds-dump-protocol.h

Setting pipeline variable containsChange=true
##vso[task.setvariable variable=containsChange;isSecret=false;isOutput=true]true
```

**Comment by am11** [comment]
Thanks for checking it, @safern.

Currently all jobs that are conditioned on `runtimetests` also have either coreclr or mono (checked with `git grep -B3 -A1 SetPathVars_runtimetests -- :/eng/pipelines`), so we are safe. If it seems implicit, we can make it explicit by adding `src/native/*` or `src/native/eventpipe/*` to `runtimetests`' `include` list.

**Comment by safern** [comment]
I think it is safe to leave it as is, I don't want to end on a world where every include path needs to be listed, unless we have a subset like the jit subset where only want to include certain paths and exclude the rest, so we implicitly include them and not exclude any paths.

**Comment by am11** [comment]
FSW test failure on OSX is unrelated to changes.

**Comment by carlossanlop** [comment]
If everything looks good, can this get merged? My PR https://github.com/dotnet/runtime/pull/61883 is currently blocked by this change.
